Make first attempt at Conways Game of Life - SamS

srsexton94 opened the pull request:
Here's my first attempt at the Hackathon project 4.15.2020

As of yet does not complete task
Stopped at rules execution

srsexton94 committed:
Make first attempt at Conways Game of Life - SamS

srsexton94 committed:
Added some logic

srsexton94 committed:
Correct typos for clarity

KytraScript commented on index.js:
```diff
@@ -60,16 +60,21 @@ const next = () => {
+      if (board[row][cell] && (neighbors == 2 || neighbors == 3)) {
+      nextBoard[row][cell] = board[row][cell]
+      // if the spot is dead but has exactly 3 neighbors, it next becomes alive
+      } else if (board[row][cell] === 0 && neighbors === 3) {
```
Disclaimer: This is just a stylistic preference, but any time you are checking to see if the value of something is === 0, you can write it as:  `if(!board[row][cell] && ...rest of code)` since 0 evaluates falsey.

srsexton94 replied:
Oh yeah that's why I chose `1` and `0` to begin with BUT theres also the chance that those values would be `undefined` which is also "falsey" - so I only wanted that if it was specifically zero

KytraScript reviewed (commented):
Nice job!

SirPotatoIV commented on index.js:
```diff
@@ -0,0 +1,80 @@
+  return Math.round(Math.random() * (max))
+}
+
+// function accepts a integer of the size of the gameboard's sides
+// it outputs a 2-dimensional array, initialized with all `0` values
+// `0` represents a dead cell, `1` represents an alive cell
```
I like that you simplified it to a 0 is dead and alive is 1. I made each one an object with a property isAlive ... this seems way cleaner.

srsexton94 replied:
Thanks! The idea was to use it in conditionals (ala `if (!cell)`) - but I forgot there could be *other* falsey values too - so it wasn't as helpful as I meant it to be haha but still works i suppose!

jcarrera94 commented on index.js:
```diff
@@ -0,0 +1,80 @@
+// function takes in an integer and outputs a random integer between 0 & input
```
I only see one file on your PR with a bunch of functions.... were you planning to implement these into a react application? it would be cool to see how you implement this file. :)

srsexton94 replied:
Yeah that was the intention!  I'm not the fastest coder yet so I wanted to just the get the functionality done - then later implement it if I can!  Thanks!

SirPotatoIV commented on index.js:
```diff
@@ -0,0 +1,80 @@
+let board = createBoard(5)
+const side = board.length
+
+// function takes in a non-zero number of cells you want to select
+// and turns that many, randomly selected indices in the sub`board` to `1`
+const selectCells = num => {
+  for (let i = 0; i <= num; i++) {
+    const row = randomInt(side)
+    const cell = randomInt(side)
+    board[row][cell] = 1
+  }
+}
```
The commenting throughout the code is great. It makes it super easy to follow. I like that you are just randomly selecting cells to make alive to begin. Seems like a good starting point before switching to having the user select the cells.

srsexton94 replied:
Thanks!

SirPotatoIV commented on index.js:
```diff
@@ -0,0 +1,80 @@
+const liveNeighbors = (r, c) => {
+  // incrementor to
+  const total = 0
+  // initializes a 2 dimensional array of all the selected squares neighbors
+  const square = [
+    [(r - 1), (c - 1)], [(r - 1), c], [(r - 1), (c + 1)],
+    [ r, (c - 1)], [r, (c + 1)],
+    [(r + 1), (c - 1)], [(r + 1), c], [(r + 1), (c + 1)]
+  ]
+  // for each truthy neighbor (ie 1) increase incrementor by 1
+  square.forEach(neighbor => {
+    if (board[neighbor[0]][neighbor[1]]) {
+      total += 1
+    }
+  })
```
Wow, starting out with making your tracking of alive vs dead as 0 and 1 makes figuring out how many neighbors are alive so much easier at this point. Had a little bit of trouble following code for creating square, but your comment does a great job giving the general idea.

srsexton94 committed:
Rewrite `liveNeighbors` to give expected output

Debug 'cannot read -1 property of undefined' variable by eliminating
undefined rows/columns before the check

carloshdelreal commented on index.js:
```diff
@@ -0,0 +1,94 @@
+
+// function takes in the row & cell placement of a square and returns number
+// of live neighbors
+const liveNeighbors = (r, c) => {
```
Try to use more readable variable names, It is very difficult to know what r or c means

srsexton94 replied:
ah yeah I was a little torn because I used `row` and `cell` in the `forEach` callbacks and didn't want to conflict with the different scopes - I can try and come up with something else though

KytraScript replied:
Here's my two-cents:  r and c are fine as they are.  We are working with a matrix/grid and you clearly stated these were rows and columns in the comments.  Single letter names aren't always a great idea, but r and c are quite intuitive here considering the given content.  It is nitpicking to criticize their use here.

srsexton94 replied:
okay - thanks for the feedback, both of yall!  I think I will leave r & c for now but I'll be sure to keep in mind more readable names in the future!
Much appreciated :)

carloshdelreal reviewed (commented):
keep going!!
